File: apps/server/agentcore/api/routes/conversations/turn_files_diff.py
"""A1+ turn files diff — read-only baseline vs live workspace."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from agentcore.api.dependencies import AuthUser, get_conversation_repo, get_message_repo
from agentcore.api.schemas.turn_files_diff import TurnFileChange, TurnFilesDiffResponse
from agentcore.core.errors import NotFoundError
from agentcore.db.repositories import ConversationRepository, MessageRepository
from agentcore.storage.protocol import SnapshotNotFound
from agentcore.workspace.turn_diff import compute_turn_files_diff

from ._helpers import _get_owned_conversation

router = APIRouter(prefix="/conversations", tags=["conversations"])


@router.get(
    "/{conversation_id}/messages/{message_id}/files/diff",
    response_model=TurnFilesDiffResponse,
)
async def get_turn_files_diff(
    conversation_id: str,
    message_id: str,
    user: AuthUser,
    conv_repo: ConversationRepository = Depends(get_conversation_repo),
    msg_repo: MessageRepository = Depends(get_message_repo),
) -> TurnFilesDiffResponse:
    """Compare the turn's baseline snapshot to the live cloud workspace (A1+).

    No baseline / snapshot missing → ``available=false`` (desktop falls back to A1
    tool-arg previews). Does not apply or restore anything.
    """
    conv = await _get_owned_conversation(conversation_id, user.user_id, conv_repo)
    msg = await msg_repo.get_by_id(message_id, conversation_id=conversation_id)
    if msg is None:
        raise NotFoundError("消息不存在")
    if msg.role != "assistant":
        raise NotFoundError("仅助手回合支持文件改动审阅")

    try:
        result = await compute_turn_files_diff(
            user_id=user.user_id,
            folder_id=conv.folder_id,
            conversation_id=conversation_id,
            message_id=message_id,
            baseline_snapshot_id=msg.baseline_snapshot_id,
        )
    except SnapshotNotFound:
        return TurnFilesDiffResponse(
            message_id=message_id,
            baseline_snapshot_id=msg.baseline_snapshot_id,
            available=False,
            data=[],
            total=0,
            added=0,
            modified=0,
            deleted=0,
        )

    rows = [
        TurnFileChange(
            path=c.path,
            change_type=c.change_type,  # type: ignore[arg-type]
            base_sha=c.base_sha,
            result_sha=c.result_sha,
            is_binary=c.is_binary,
            content=c.content,
            size_bytes=c.size_bytes,
            base_content=c.base_content,
        )
        for c in result.changes
    ]
    added = sum(1 for r in rows if r.change_type == "added")
    modified = sum(1 for r in rows if r.change_type == "modified")
    deleted = sum(1 for r in rows if r.change_type == "deleted")
    return TurnFilesDiffResponse(
        message_id=message_id,
        baseline_snapshot_id=result.baseline_snapshot_id,
        available=result.available,
        data=rows,
        total=len(rows),
        added=added,
        modified=modified,
        deleted=deleted,
    )

File: apps/desktop/src/renderer/services/turnFilesDiff.ts
/**
 * A1+ 回合文件真 diff —— 云端 REST / 本机 sidecar RPC。
 * 无基线 / 失败时 ``available=false``，桌面降级工具参数预览。
 */

import { api } from "@/services/api";

export type TurnChangeType = "added" | "modified" | "deleted";

export interface TurnFileChange {
  path: string;
  changeType: TurnChangeType;
  baseSha: string | null;
  resultSha: string | null;
  isBinary: boolean;
  content: string | null;
  sizeBytes: number;
  baseContent: string | null;
}

export interface TurnFilesDiff {
  messageId: string;
  baselineSnapshotId: string | null;
  available: boolean;
  changes: TurnFileChange[];
  total: number;
  added: number;
  modified: number;
  deleted: number;
}

export interface LocalTurnFilesTarget {
  rootId: string;
  subpath?: string;
}

interface WireChange {
  path: string;
  change_type: TurnChangeType;
  base_sha: string | null;
  result_sha: string | null;
  is_binary: boolean;
  content: string | null;
  size_bytes: number;
  base_content?: string | null;
}

interface WireDiff {
  message_id: string;
  baseline_snapshot_id: string | null;
  available: boolean;
  data: WireChange[];
  total: number;
  added: number;
  modified: number;
  deleted: number;
}

function mapChange(c: WireChange): TurnFileChange {
  return {
    path: c.path,
    changeType: c.change_type,
    baseSha: c.base_sha,
    resultSha: c.result_sha,
    isBinary: c.is_binary,
    content: c.content,
    sizeBytes: c.size_bytes,
    baseContent: c.base_content ?? null,
  };
}

function fromWire(raw: WireDiff): TurnFilesDiff {
  return {
    messageId: raw.message_id,
    baselineSnapshotId: raw.baseline_snapshot_id,
    available: raw.available,
    changes: (raw.data ?? []).map(mapChange),
    total: raw.total,
    added: raw.added,
    modified: raw.modified,
    deleted: raw.deleted,
  };
}

/** Cloud path: ``GET …/messages/{id}/files/diff``. */
export async function getTurnFilesDiff(
  conversationId: string,
  messageId: string,
): Promise<TurnFilesDiff> {
  const raw = await api.get<WireDiff>(
    `/v1/conversations/${conversationId}/messages/${messageId}/files/diff`,
  );
  return fromWire(raw);
}

/** Local path: sidecar RPC over `.agentcore/baselines/{messageId}.zip` (no cloud). */
export async function getLocalTurnFilesDiff(
  target: LocalTurnFilesTarget,
  messageId: string,
): Promise<TurnFilesDiff> {
  const raw = await window.sidecarApi.turnFilesDiff({
    rootId: target.rootId,
    subpath: target.subpath,
    messageId,
  });
  return fromWire(raw);
}

/** Local A2′ restore via sidecar unzip (never cloud restoreSnapshot). */
export async function restoreLocalTurnBaseline(
  target: LocalTurnFilesTarget,
  snapshotId: string,
): Promise<void> {
  await window.sidecarApi.restoreTurnBaseline({
    rootId: target.rootId,
    subpath: target.subpath,
    snapshotId,
  });
}
